### PR TITLE
Extract out mockable interfaces for KubeClient

### PR DESCRIPTION
**Pull request by adammw**
### Description

Includes changes from #4388 (ideally review and merge separately) to reduce merge conflicts while developing.

This change adds the plumbing needed to be able to mock out the kubeclient in `ClusterProvider` with a fake one, for example in testing cluster build tasks.

ClusterProvider's `Provider` field has been split into separate `AWSProvider` (for the AWS client wrapper) and `KubeProvider` for the (kubernetes client wrapper). The previous `KubeProvider` interface which was a partial interface for ClusterProvider has been replaced with ClusterProvider itself, which is now an interface (and the concrete implementation called ClusterProviderImpl - other naming suggestions welcome). It also updates a lot of methods/structs to accept interface pointers allowing the use of mocked client implementations. 

### Checklist
- [x] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [ ] Manually tested
- [ ] Made sure the title of the PR is a good description that can go into the release notes
- [ ] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [x] Backfilled missing tests for code in same general area :tada:
- [x] Refactored something and made the world a better place :star2:

### CODE REVIEW

**Comment by adammw** [comment]
Tests seem to be passing locally. Not sure what the next steps here are, I got a bit carried away in how many changes I ended up making cleaning up some of the confusingly named structs/interfaces.

**Comment by adammw** [comment]
@aclevername @cPu1 @Skarlso could you take a look if this is a set of changes the project would benefit from, and how to best proceed? 
cc @uthark @grosser as well for other golang tips

**Comment by aclevername** [comment]
Thanks for opening the PR @adammw - I'm a bit swamped right now but will try to get to this ASAP :smile: I've requested review from the rest of the team in the meantime

**Comment by aclevername** [comment]
> @aclevername @cPu1 @Skarlso could you take a look if this is a set of changes the project would benefit from, and how to best proceed? cc @uthark @grosser as well for other golang tips

Adding @Himangini @nikimanoledaki

File: pkg/eks/api.go [hunk 1]
```diff
@@ -44,29 +44,47 @@ import (
...
-	"github.com/weaveworks/eksctl/pkg/kubernetes"
+	iamoidc "github.com/weaveworks/eksctl/pkg/iam/oidc"
 	kubewrapper "github.com/weaveworks/eksctl/pkg/kubernetes"
+	"github.com/weaveworks/eksctl/pkg/utils/tasks"
 	"github.com/weaveworks/eksctl/pkg/version"
 )
 
-// ClusterProvider stores information about the cluster
-type ClusterProvider struct {
+// ClusterProviderImpl stores information about the cluster
```
Comment (Skarlso) [inline pkg/eks/api.go:53]: Please don't call them `Impl` and `...Interface`. :)

File: pkg/testutils/client.go [hunk 2]
```diff
@@ -300,3 +303,28 @@ func (c *FakeRawClient) ClearUpdated() {
 		delete(c.Collection.updated, k)
 	}
 }
+
+func (c *FakeRawClient) ServerVersion() (string, error) {
+	return getServerVersion(c.ClientSet().Discovery())
+}
+
+func (c *FakeRawClient) CreateOrReplace(manifest []byte, plan bool) error {
+	panic("not implemented") // TODO
```
Comment (Skarlso) [inline pkg/testutils/client.go:312]: TODO

File: pkg/testutils/mockprovider/mock_kube_provider.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
+package mockprovider
```
Comment (Skarlso) [inline pkg/testutils/mockprovider/mock_kube_provider.go:1]: Can we use `go:generate` for this?

File: pkg/kubernetes/client.go [hunk 4]
```diff
@@ -67,6 +68,12 @@ func (g *CallbackClientSet) ClientSet() (Interface, error) {
 	return g.Callback()
 }
 
+// ClientInterface defines the high level client used by eks pkg
+type ClientInterface interface {
+	NewClientSet() (Interface, error)
+	Config() *clientcmdapi.Config
```
Comment (Skarlso) [inline pkg/kubernetes/client.go:74]: Do I see correctly that this isn't call by anything anywhere?

Comment (Skarlso) [reply pkg/kubernetes/client.go:74]: What is the purpose of this `Config` method?

File: pkg/eks/api.go (hunk 1 above)
Comment (adammw) [reply pkg/eks/api.go:53]: Open to alternative names, I just couldn't think of any better ones at the time.

File: pkg/kubernetes/client.go (hunk 4 above)
Comment (adammw) [reply pkg/kubernetes/client.go:74]: Seems to be used in the cluster creation gitops flow... would love to get rid of this layer of abstraction and just have dedicated NewClientSet / Config methods in the KubeProvider interface instead.

Comment (adammw) [reply pkg/kubernetes/client.go:74]: Seems to be a similar problem with RawClientInterface which was pre-existing before the refactors in this PR, although that has a lot more methods on it (NewRawResource, ServerVersion, CreateOrReplace)

File: pkg/testutils/mockprovider/mock_kube_provider.go (hunk 3 above)
Comment (adammw) [reply pkg/testutils/mockprovider/mock_kube_provider.go:1]: This provides an implementation that returns a statically created client, not just stubbed fields. If it should go elsewhere (or be named something other than 'mock' to distinguish between them) I'm all ears.

File: pkg/eks/api.go (hunk 1 above)
Comment (Skarlso) [reply pkg/eks/api.go:56]: Will think of something. :)

File: pkg/kubernetes/client.go (hunk 4 above)
Comment (Skarlso) [reply pkg/kubernetes/client.go:74]: Ah yes. Somehow I missed that usage.

**Comment by adammw** [comment]
I might park this change as it's grown wildly in scope, and still doesn't accomplish what I set out to achieve which was to be able to unit test/stub/mock a cluster build start to finish. (even with the kube client mocked there's now OIDC provider to contend with)

**Comment by adammw** [comment]
I might park this change as it's grown wildly in scope, and still doesn't accomplish what I set out to achieve which was to be able to unit test/stub/mock a cluster build start to finish. (even with the kube client mocked there's now OIDC provider to contend with)

**Comment by Himangini** [comment]
> I might park this change as it's grown wildly in scope, and still doesn't accomplish what I set out to achieve which was to be able to unit test/stub/mock a cluster build start to finish. (even with the kube client mocked there's now OIDC provider to contend with)

Thanks, @adammw I am removing us from the reviews list. Please tag us again when this PR is ready to be reviewed.

**Comment by Skarlso** [comment]
Closing as this was superseded by various refactors and we'll add create unit tests using a mock server in this ticket: https://github.com/weaveworks/eksctl/issues/5015

**Pull request closed without merging**